cpp/src/strings/convert/convert_urls.cu

```diff
@@ -222,30 +331,87 @@ std::unique_ptr<column> url_decode(
+  size_type chars_bytes = chars_end - chars_start;
+
+  url_decode_escape_detector esc_detector{device_span<char const>(d_in_chars, chars_end)};
+
+  // Count the number of URL escape sequences across all strings, ignoring string boundaries.
+  // This may count more sequences than actually are there since string boundaries are ignored.
+  size_type esc_count = thrust::count_if(rmm::exec_policy(stream),
+                                         thrust::make_counting_iterator<size_type>(chars_start),
+                                         thrust::make_counting_iterator<size_type>(chars_end),
+                                         esc_detector);
+
+  if (esc_count == 0) {
+    // nothing to replace, so just copy the input column
+    return std::make_unique<cudf::column>(strings.parent());
+  }
+
+  // create a vector of the potential escape sequence positions
+  rmm::device_uvector<size_type> esc_positions(esc_count, stream);
+  auto d_esc_positions = esc_positions.data();
+  thrust::copy_if(rmm::exec_policy(stream),
+                  thrust::make_counting_iterator<size_t>(chars_start),
+                  thrust::make_counting_iterator<size_t>(chars_end),
+                  d_esc_positions,
+                  esc_detector);
+
+  // In-place remove any escape positions that crossed string boundaries.
+  device_span<int32_t const> d_offsets_span(d_offsets, offset_count);
+  auto esc_pos_end = thrust::remove_if(rmm::exec_policy(stream),
+                                       d_esc_positions,
+                                       d_esc_positions + esc_count,
+                                       url_decode_esc_position_filter{d_offsets_span});
+
+  // update count in case any were filtered
+  esc_count = esc_pos_end - d_esc_positions;
+  if (esc_count == 0) {
+    // nothing to replace, so just copy the input column
+    return std::make_unique<cudf::column>(strings.parent());
+  }
+
+  device_span<size_type const> d_esc_positions_span(d_esc_positions, esc_count);
 
-  // copy null mask
-  rmm::device_buffer null_mask = cudf::detail::copy_bitmask(strings.parent(), stream, mr);
   // build offsets column
-  auto offsets_transformer_itr = thrust::make_transform_iterator(
-    thrust::make_counting_iterator<size_type>(0), url_decoder_fn{d_strings});
-  auto offsets_column = make_offsets_child_column(
-    offsets_transformer_itr, offsets_transformer_itr + strings_count, stream, mr);
-  auto d_offsets = offsets_column->view().data<int32_t>();
+  auto offsets_column = make_numeric_column(
+    data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+  auto offsets_view = offsets_column->mutable_view();
+  thrust::transform(rmm::exec_policy(stream),
+                    d_offsets_span.begin(),
+                    d_offsets_span.end(),
+                    offsets_view.begin<size_type>(),
+                    url_decode_offsets_updater{d_esc_positions_span, chars_start});
```

Comment:
This should work, if you want to avoid having a variable just to pass it as a parameter.
```suggestion
                    url_decode_offsets_updater{{d_esc_positions, esc_count}, chars_start});
```

Reply:
`d_esc_positions_span` is used as an argument more than once which is why I decided to make it a local variable.  If you'd rather see the explicit span args in those places instead I'm happy to change it.

Reply:
It's fine as is then, I missed the reuse.